contracts/GenesisManager.sol

```diff
@@ -0,0 +1,268 @@
+
+        addVestingGrant(_receiver, _amount, _timeToCliff, _vestingDuration);
+    }
+
+    /**
+     * @dev Add a grant for tokens with a vesting schedule. An internal helper function used by addTeamGrant and addInvestorGrant
+     * @param _receiver Grant receiver
+     * @param _amount Amount of tokens included in the grant
+     * @param _timeToCliff Seconds until the vesting cliff
+     * @param _vestingDuration Seconds starting from the vesting cliff until the end of the vesting schedule
+     */
+    function addVestingGrant(
+        address _receiver,
+        uint256 _amount,
+        uint256 _timeToCliff,
+        uint256 _vestingDuration
+    )
+        internal
+    {
+        // Receiver must not have already received a grant with a vesting schedule
+        require(vestingHolders[_receiver] == address(0));
+
+        // The grant's vesting schedule starts when the token distribution ends
+        uint256 startTime = tokenDistribution.getEndTime();
+        // Create a vesting holder contract to act as the holder of the grant's tokens
+        // Note: the vesting grant is revokable
+        TokenVesting holder = new TokenVesting(_receiver, startTime, _timeToCliff, _vestingDuration, true);
+        vestingHolders[_receiver] = holder;
+
+        // Transfer ownership of the vesting holder to the owner of this contract
+        // giving the owner of this contract to revoke the grant
+        holder.transferOwnership(owner);
+
+        token.transfer(holder, _amount);
+    }
+
+    /**
+     * @dev Add a community grant for tokens that are locked until a predetermined time in the future
+     * @param _receiver Grant receiver address
+     * @param _amount Amount of tokens included in the grant
+     */
+    function addCommunityGrant(
+        address _receiver,
+        uint256 _amount
+    )
+        external
+        onlyOwner
+        atStage(Stages.GenesisStart)
+    {
+        uint256 updatedGrantsAmount = communityGrantsAmount.add(_amount);
+        // Amount of tokens included in investor grants cannot exceed the community supply during genesis
+        require(updatedGrantsAmount <= communitySupply);
+
+        communityGrantsAmount = updatedGrantsAmount;
+
+        // Receiver must not have already received a grant with timelocked tokens
+        require(timeLockedHolders[_receiver] == address(0));
+
+        // The grant's tokens are timelocked until the token distribution ends
+        uint256 releaseTime = tokenDistribution.getEndTime();
+        // Create a timelocked holder contract to act as the holder of the grant's tokens
+        TokenTimelock holder = new TokenTimelock(token, _receiver, releaseTime);
+        timeLockedHolders[_receiver] = holder;
+
+        token.transfer(holder, _amount);
+    }
+
+    /**
+     * @dev End genesis
+     */
+    function end() external onlyOwner atStage(Stages.GenesisStart) {
+        // Token distribution must be over
+        require(tokenDistribution.isOver());
```

Comment:
Up for discussion whether this returns true when the registration period is over or the actual initial distribution is over.